### PR TITLE
Proposing a new branching and merging model

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

This PR updates our READMEs to detail a proposed update to our branching and merging model that @Martinsos and I chatted about today. The current "deploy everything from `main`" model suffers from:
- Inability to deploy website content independent of changes already in `main`.
  - This means we must delay merging code with corresponding doc updates into `main` if we want to be able to publish a new blog post before we cut a new release, for example, as otherwise, our docs won't match the current Wasp release.
- We don't have a good way to handle hotfixes.
  - Many times, `main` may have breaking changes staged to go out next. If we had to make a hotfix for the current release we could not use `main` since building a new release from there could break the current release.

This model will likely have to evolve over time. For example, we probably want to be able to merge hotfixes into multiple prior releases, not just the current. But this idea should improve our current setup in the meantime.

What does everyone think? I am especially curious if there are better ways to handle the divergence of `release` and `main`.

_UPDATE: The more I think about it, the more I think a [gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) model may make sense for us long-term. We could make our releases branches we could then prepare more methodically and keep around to easily hotfix as well. Note, this is not part of this idea and I know it deviates from the popular trunk based used by many SaaS apps today, but I think our ultimate ideal flow is necessarily different._

# TODOs to roll this out:
- [ ] create a new branch `release` from `main`
- [ ] retarget all website-only PRs from `main` to `release`
- [ ] test a website deploy from the `release` branch
- [x] test the git backmerge process from `release` into `main` and how that impacts future merges from `main` to `release` and the history locally (Martin + Shayne did)
- [ ] add CI to `release`? ensure it all still plays nicely with tags like `v`
- [ ] test the Wasp release process to use `release` instead of `main`

### CODE REVIEW

**Comment by Martinsos** [comment]
This makes sense to me!
Yup, trunk development is cool and we are doing it in a way, but I don't think it is enough for us since we have specific versions of Wasp that we want to release only at specific times -> we cant release Wasp multiple times per day :D.

So we can still use "trunk" philosophy for our "edge" version of Wasp, that we are developing on `main`, in the way that we can ensure PR towards it are relatively small and we try to merge them quickly, and we have CI running on it, that runs tests all the time and ensures it builds.

But, we need to have a way to also modify existing previous versions of Wasp, which include previous releases (e.g. 0.2, 0.3) and it also in a sense includes Wasp web (docs + blog).
For now we care only about the latest released version of Wasp, which simplifies things, as we can just have `release` branch to track stuff that is released -> that means waspc and web. Ok, maybe also other stuff like vscode extension at some point hm.

I am just thinking out loud here, but yeah I agree with everything, including the part that it is not perfect solution, but it feels like a good enough start and we can figure out something better once we need it.

As for backporting -> hm, how about we just push stuff on `release` when it is stuff like hotfixes, blog posts, doc hotfixes, that we want to have out there right now, ..., and we push stuff on `main` when it is new stuff that we are ok with coming out with the next possibly breaking change release. And then once `main` is ready to be released, we merge it into `release`, resolve the conflicts and we are good? So in theory we don't need to do any backporting from `release` to `main`, instead we will just be merging `main` into `release` every so and so. If we really need smth from `release`, well we can cherry-pick it or copy it or smth. Maybe merge if we really have to but that will mess up commit history, I would avoid that.

Ok, for code hotfixes we actually need to make new patch release, but ok that makes sense.

Yup, sounds all good to me! I will review the code now.

**Comment by Martinsos** [comment]
Oops I was being silly up there -> of course we need to backport from `release` into `main`, otherwise `main` would forever keep missing those updates. Hmmmmm.
Ok so I guess when we merge `main` into `release`, we also do the opposite thing, and make sure both are equal at that point? Yup, no nice git history, but who cares.

**Comment by Martinsos** [comment]
We could possibly use `git merge --squash` when backporting from `release` into `main`, I think that should ensure the git history in `main` remains linear, although I am not 100% sure.

File: waspc/README.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ If you would like to make your first contribution, here is a handy checklist we
 - [ ] Join [Discord](https://discord.gg/rzdnErX) and say hi :)!
 - [ ] Pick an issue [labeled with "good first issue"](https://github.com/wasp-lang/wasp/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) and let us know you would like to work on it - ideally immediatelly propose a plan of action and ask questions.
       If you can't find a suitable issue for you, reach out to us on Discord and we can try to find smth for you together.
-- [ ] Make a PR and have it accepted! Check [Typical workflow](#typical-development-workflow) for guidance, and consult [Codebase overview](#codebase-overview) for more details on how Wasp compiler works internally.
+- [ ] Make a PR targeting `main` and have it accepted! Check [Typical workflow](#typical-development-workflow) for guidance, and consult [Codebase overview](#codebase-overview) for more details on how Wasp compiler works internally.
```
Comment (Martinsos) [inline waspc/README.md:16]: Maybe they want to target `release` if they are fixing docs, or a bug?
But ok, maybe it is best they target `main` and we can correct them if needed.

File: waspc/README.md [hunk 2]
```diff
@@ -316,6 +316,21 @@ NOTE: When you run it for the first time it might take a while (~10 minutes) for
...
 We use [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/) convention when creating commits.
 
+## Branching and merging strategy
+This repo contains both the source code that makes up a Wasp release (under `waspc`), as well as our website containing documentation and blog posts (under `web`). In order to facilitate the development of Wasp code while still allowing for website updates or hotfixes of the current release, we have decided on the following minimal branching strategy.
+
+All Wasp development should be done on feature branches. They form the basis of PRs that will target one of the two following branches:
+
+- `main`: this branch contains all the actively developed new features and corresponding documentation updates. Some of these things may not yet be released, but anything merged into `main` should be in a release-ready state.
+  - This is the default branch to target for any Wasp feature branches.
+- `release`: this branch contains the source code of current/latest Wasp release, as well as the documentation and blog posts currently visible on the website.
```
Comment (Martinsos) [inline waspc/README.md:326]:
```suggestion
- `release`: this branch contains the source code of current/latest Wasp release, as well as the documentation and blog posts currently published and therefore visible on the website.
```

File: waspc/README.md [hunk 3]
```diff
@@ -316,6 +316,21 @@ NOTE: When you run it for the first time it might take a while (~10 minutes) for
...
+  - When we do a full release, we first backmerge `release` into `main` to pick up any diverging changes (see below). We then merge `main` into `release` to get all the new changes. We can now create a new Wasp release as well as deploy the website from the `release` branch.
```
Comment (Martinsos) [inline waspc/README.md:327]:
```suggestion
  - When doing a full release, which means making a new release based on what we have currently on `main`, we do the following:
    1. Update `release` branch by merging `main` into it. There might be conflicts, since `release` might have hotfixes and blog posts that `main` doesn't have, but they shouldn't be too hard to fix. Once `release` is updated, you can create a new waspc release from it, as well as deploy the website from it.
    2. To keep `main` up to date with any changes in `release`, you should now also update `main` to reflect the new `release`. There will be no conflicts since we already resolved all of them in the previous step.
```

Comment (Martinsos) [reply waspc/README.md:327]: I tried to simplify language a bit, since I found it a bit too technical for me!

And I believe you can then remove the explanation at the bottom, under NOTE.

File: waspc/README.md [hunk 4]
```diff
@@ -335,13 +350,16 @@ If it happens just once every so it is probably nothing to worry about. If it ha
...
+- Merge `main` into `release`.
+- Make sure you are on `release` and then run `./new-release 0.x.y.z`.
   - This will automatically create a new commit for updating the version in waspc.cabal, tag it, and push it all.
 - Wait for CI to finish & succeed for the new tag.
   - This will automatically create a new draft release.
 - Find new draft release here: https://github.com/wasp-lang/wasp/releases and edit it with your release notes.
 - Publish the draft release when ready.
-- Publish new [docs](/web#deployment).
+- Publish new [docs](/web#deployment) from the `release` branch as well.
 - Announce new release in Discord.
```
Comment (Martinsos) [inline waspc/README.md:363]:
```suggestion
- Make sure your `release` and `main` branches are up to date with the remote (run `git fetch`).
- Update `release` with the `main` by checking out `release` and running `git merge main`. Resolve any conflicts.
- Make sure you are on `release` and then run `./new-release 0.x.y.z`.
  - This will automatically create a new commit for updating the version in waspc.cabal, tag it, and push it all.
- Wait for CI to finish & succeed for the new tag.
  - This will automatically create a new draft release.
- Find new draft release here: https://github.com/wasp-lang/wasp/releases and edit it with your release notes.
- Publish the draft release when ready.
- Publish new [docs](/web#deployment) from the `release` branch as well.
- Announce new release in Discord.
- Update `main` with the `release` by checking out `main` and running `git merge --squash release`. There should be no conflicts.
```

Maybe we should not do `--squash` in that last merge, I am not sure. `--squash` will ensure history is kept linear in `main`. But it will squash the work done on `release`. I am not sure if this is better or it is better to have merges in two directions.

File: waspc/README.md [hunk 5]
```diff
@@ -316,6 +316,21 @@ NOTE: When you run it for the first time it might take a while (~10 minutes) for
...
+
+All Wasp development should be done on feature branches. They form the basis of PRs that will target one of the two following branches:
+
+- `main`: this branch contains all the actively developed new features and corresponding documentation updates. Some of these things may not yet be released, but anything merged into `main` should be in a release-ready state.
+  - This is the default branch to target for any Wasp feature branches.
+- `release`: this branch contains the source code of current/latest Wasp release, as well as the documentation and blog posts currently visible on the website.
+  - When we do a full release, we first backmerge `release` into `main` to pick up any diverging changes (see below). We then merge `main` into `release` to get all the new changes. We can now create a new Wasp release as well as deploy the website from the `release` branch.
+  - How can `release` diverge from `main`?
+    - If you have a website-only update (like a blog post or documentation fix) that does _not_ depend on a Wasp release, you may open your PR against `release`.
+    - If there is a hotfix required for the _current_ Wasp release, you may open your PR against `release`.
```
Comment (Martinsos) [inline waspc/README.md:330]:
```suggestion
How do I know where I want to target my PR, to `release` or `main`?
  - If you have a change that you want to publish right now or very soon, certainly earlier than waiting till `main` is ready for publishing, then you want to target `release`. This could be website content update, new blog post, documentation (hot)fix, compiler hotfix that we need to release quickly via a new patch version, ... .
  - If you have a change that is not urgent and can wait until the next "normal" Wasp release is published, then target `main`. These are new features, refactorings, docs accompanying new features, ... .
  - TLDR;
    - `release` is for changes to the already published stuff (the present).
    - `main` is for changes to the to-be-published stuff (the future).
```

File: waspc/README.md [hunk 6]
```diff
@@ -335,13 +350,16 @@ If it happens just once every so it is probably nothing to worry about. If it ha
 ### Typical Release Process
 - Update ChangeLog.md with release notes and open an PR for feedback.
 - After approval, squash and merge PR for ChangeLog.md into `main`.
-- Make sure you are on `main` and up to date locally :D and then run `./new-release 0.x.y.z`.
+- Update your local repository state to have all remote changes.
```
Comment (Martinsos) [inline waspc/README.md:357]:
```suggestion
- Update your local repository state to have all remote changes (`git fetch`).
```

File: waspc/README.md [hunk 7]
```diff
@@ -335,13 +350,16 @@ If it happens just once every so it is probably nothing to worry about. If it ha
...
+- Backmerge `release` into `main`.
+- Merge `main` into `release`.
```
Comment (Martinsos) [inline waspc/README.md]:
```suggestion
- Update `main` to contain changes from `release` by running `git merge release` while on the `main` branch. Resolve any conflicts.
- Fast-forward `release` to this new, updated `main` by running `git merge main` while on the `release` branch.
```

File: waspc/README.md (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/README.md:16]: Yeah, I was trying to just be safe with the most common case as the suggestion, hoping we can correct if needed or they look at the details below. I can add a link to the new section here as well if that may help?

**Comment by Martinsos** [comment]
I wrongly suggested we merge main into release first, and then release into main, instead of doing the opposite. I believe I fixed it now, but if some comment is a bit weird / silly still, it might be due to that so pls compensate for that!

File: waspc/README.md (hunk 1 above)
Comment (Martinsos) [reply waspc/README.md]: I think this is fine!

**Comment by sodic** [comment]
Everything makes sense to me! I won't go into editing suggestions, I see Martin already covered the entire instructions (and then some :)).

Anyway, one thing we might want to consider is separating blog posts and the website into a different repository. We currently can't release a new blog post without releasing a new version of Wasp (or, assuming this new system, by abusing the release branch).

When I say blog and docs, I mean stuff like blog posts, tutorials... I don't mean documentation (i.e., API reference). Martin suggested that it makes sense to keep the docs coupled with the code. I see how it makes sense, but it might be tricky integrating it with the rest of the website.

In any case, this change is fine for now. No need to do anything else as far as I'm concerned :)

**Review by sodic — APPROVED** [review]

**Comment by shayneczyzewski** [comment]
Thanks so much for the review, suggestions, and back and forth on Discord @Martinsos !

Thanks for your comments and review/approval @sodic as well. I did have the same thought about pulling out blog/non-doc website stuff, but I think we figured we may be able to solve the hotfix current release problem and that one in the same way for now. But agree, over time, it may make sense. Martin also noted we may want to update blog post example code as things change too, so having them together is nice (just like the doc case).

Also, just so you are aware, we did some local testing and confirmed that merging `release` into `main` first to backport any deviations, and then merging `main` to `release` resulted in the least noisy tree with the most semantically correct merge commit messages, like so (note in my testing `web` was `release` and `main` was `master`:
Screen Shot 2022-08-17 at 9 21 42 AM

General question- are you aware of any downsides or git alternatives to this approach to get these branches equivalent again we should consider?

**Comment by Martinsos** [comment]
> Everything makes sense to me! I won't go into editing suggestions, I see Martin already covered the entire instructions (and then some :)).
> 
> Anyway, one thing we might want to consider is separating blog posts and the website into a different repository. We currently can't release a new blog post without releasing a new version of Wasp (or, assuming this new system, by abusing the release branch).
> 
> When I say blog and docs, I mean stuff like blog posts, tutorials... I don't mean documentation (i.e., API reference). Martin suggested that it makes sense to keep the docs coupled with the code. I see how it makes sense, but it might be tricky integrating it with the rest of the website.
> 
> In any case, this change is fine for now. No need to do anything else as far as I'm concerned :)

I like @sodic 's suggestion to separate webpage and blog posts! It is nice to keep docs next to the code, but we need to draw a line where the docs end and where the rest of the content starts. For example, docs could be API reference + quick guide + couple of tutorials, but the rest of the blog posts + webpage could be separate repo, and possibly also example apps could be extracted. I will create an issue to address this! But for now we don't have to worry about it.

File: waspc/README.md (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/README.md]: Just added that link 👍🏻

**Comment by sodic** [comment]
One thing to consider: the history will definitely get messier when we start applying those hotfixes frequently, which is a shame because we have a pretty nice git history, but is not really that big of a deal I think.

One alternative approach that comes to mind goes like this (but I don't like it more than I like this one):
1. We release version 1.5
2. We tag the commit with `v1.5`
3. We continue working and pushing stuff to `main`
4. We realize we need to make a hotfix for the latest version
5. We checkout `v1.5`, branch out, implement the hotfix, and then release that branch as `v1.5.1` and merge it back into `main`
6. All future hotfixes for branch `v1.5` go on the newly created branch

File: waspc/README.md (hunk 3 above)
Comment (shayneczyzewski) [reply waspc/README.md]: Updated with your suggestion (with one tweak to clarify the how: "2. Update `release` branch to this new `main` by merging `main` into it.") Removed NOTE and references to backmerge. Thanks! 👍🏻

**Review by Martinsos — APPROVED** [review]

File: waspc/README.md (hunk 6 above)
Comment (shayneczyzewski) [reply waspc/README.md]: Updated, thanks!

File: waspc/README.md (hunk 7 above)
Comment (shayneczyzewski) [reply waspc/README.md]: Updated!

**Comment by shayneczyzewski** [comment]
> One thing to consider: the history will definitely get messier when we start applying those hotfixes frequently, which is a shame because we have a pretty nice git history, but is not really that big of a deal I think.
> 
> One alternative approach that comes to mind goes like this (but I don't like it more than I like this one):
> 
> 1. We release version 1.5
> 2. We tag the commit with `v1.5`
> 3. We continue working and pushing stuff to `main`
> 4. We realize we need to make a hotfix for the latest version
> 5. We checkout `v1.5`, branch out, implement the hotfix, and then release that branch as `v1.5.1` and merge it back into `main`
> 6. All future hotfixes for branch `v1.5` go on the newly created branch

Good points, I think over time we will need to evolve this setup for longterm hotfixes. We can maybe try to evolve this proposed process as those needs arise in future.
